### ISSUE TITLE
User Claims missing from Identity Endpoints breaks Authorization

### ISSUE THREAD

**Issue by augustevn** [dotnet/aspnetcore#52142]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

I did some digging, it seems like the user Claims are no longer returned from `/manage/info` from the `.MapIdentityApi<User>()`.

My custom authorization was depending on getting those claims.
This change broke my app, worked in RC2 not in RTM. Or is this intentional?

`IdentityApiEndpointRouteBuilderExtensions.cs`, line 455:

```

    private static async Task<InfoResponse> CreateInfoResponseAsync<TUser>(TUser user, UserManager<TUser> userManager)
        where TUser : class
    {
        return new()
        {
            Email = await userManager.GetEmailAsync(user) ?? throw new NotSupportedException("Users must have an email."),
            IsEmailConfirmed = await userManager.IsEmailConfirmedAsync(user),
        };
    }

```

### Expected Behavior

Return the user's Claims on authenticated `GET` call to `/manage/info` on the Identity Endpoints.

### Steps To Reproduce

Upgrade from .NET 8 RC2 to .NET 8 RTM, implement the `.MapIdentityApi<User>()`, spot the 7 differences.

### Exceptions (if any)

/

### .NET Version

8.0.100

### Anything else?

Possibly related to Blazor Identity UI issues: https://github.com/dotnet/aspnetcore/issues/52063
Of which it likely affects authorization.

**Comment by Tratcher** [comment]
Yes, this was by design: https://github.com/dotnet/aspnetcore/pull/51177

**Comment by augustevn** [comment]
Oh thanks, I'll try to get it from the HttpContextAccessor than?

**Comment by augustevn** [comment]
@Tratcher How do you suggest Identity API consumers to handle authorization than, if they can't access the claims? For stateless bearer token auth.

**Comment by Tratcher** [comment]
Authorization must always be done on the server side (where you have access to the Claims in the User). Anything done on the client is just for convenience. You can also expose your own endpoints for the API to check for specific permissions.

**Comment by augustevn** [comment]
@Tratcher I understand but how can a `<AuthorizeView Roles="Administrator" />` work if we can't get the roles from the endpoints nor from the token?

I'm afraid this is another reason not to use the Identity Endpoints... Client-side SPA's are likely the only ones using them. If we need to build multiple custom endpoints ourselves, we might as well build the other 3 endpoints ourselves as well.

I appreciate the effort but I don't see this attempt end well. It's lacking customizability due to DTO usage and now this...

**Comment by augustevn** [comment]
Let me try again, I think you broke the Blazor WASM way of informing our `AuthenticationStateProvider` of our claims. At least you broke the seamless integration of that with the Identity Endpoints.

Your action of removing the claims from the GET `/manage/info` response, now broke this. I think that is still a better practice than having to verify and decode the bearer token to get the claims? Since that implies we'd need secrets in our front-end?

Take the below `CustomAuthenticationStateProvider`:

``` csharp

public class CustomAuthenticationStateProvider : AuthenticationStateProvider
{
    public override Task<AuthenticationState> GetAuthenticationStateAsync()
    {
        var emptyAuthState = ConstructEmptyAuthState();
        return Task.FromResult(emptyAuthState);
    }

    public void Authenticate(UserInfoResponse currentUser)
    {
        var identity = new ClaimsIdentity(
            currentUser.Claims.Select(c => new Claim(c.Key, c.Value.ToString())),
            "auth");
        
        var user = new ClaimsPrincipal(identity);
        var authState = new AuthenticationState(user);
        
        NotifyAuthenticationStateChanged(Task.FromResult(authState));
    }

    public void UnsetAuthenticationState()
    {
        var emptyAuthState = ConstructEmptyAuthState();
        
        NotifyAuthenticationStateChanged(Task.FromResult(emptyAuthState));
    }

    private static AuthenticationState ConstructEmptyAuthState()
    {
        var identity = new ClaimsIdentity();
        var user = new ClaimsPrincipal(identity);
        var authState = new AuthenticationState(user);

        return authState;
    }
}

```

**Comment by julioct** [comment]
I also stumbled into this today. It didn't break my app, since I was just trying to integrate my Blazor WASM app with the new identity endpoints as described in [this official doc](https://learn.microsoft.com/aspnet/core/blazor/security/webassembly/standalone-with-identity).

However, after completing the work, I realized the identity endpoints are not useful other than to register and authenticate a user.

And, the first thing you want to do after authenticating the user is check which permissions he has to decide how much to allow him to do on both the server and client side.

Without the claims, you can't tell the most basic thing: is he an admin? Can he modify things on my site?

I understand that this is not OIDC, so I would never be able to set scopes, which is the next thing I would want. But I must be able to at least check basic claims like the role the user belongs to.

To unblock, I implemented my own Identity endpoint group by extending the built-in ones:

```csharp
public static class IdentityEndpoints
{
    public static RouteGroupBuilder MapIdentityEndpoints(this IEndpointRouteBuilder routes)
    {
        var group = routes.MapGroup("/identity")
                          .WithTags("Identity");

        group.MapIdentityApi<GameStoreUser>();

        group.MapPost("/logout", async (ClaimsPrincipal user, SignInManager<GameStoreUser> signInManager) =>
        {
            await signInManager.SignOutAsync();
            return TypedResults.Ok();
        });

        group.MapGet("/manage/infoWithClaims", async Task<Results<Ok<InfoWithClaimsResponse>, ValidationProblem, NotFound>>
            (ClaimsPrincipal claimsPrincipal, UserManager<GameStoreUser> userManager) =>
        {
            if (await userManager.GetUserAsync(claimsPrincipal) is not { } user)
            {
                return TypedResults.NotFound();
            }

            var email = await userManager.GetEmailAsync(user) ?? throw new NotSupportedException("Users must have an email.");

            var response = new InfoWithClaimsResponse(
                email,
                await userManager.IsEmailConfirmedAsync(user),
                claimsPrincipal.Claims.ToDictionary(c => c.Type, c => c.Value));

            return TypedResults.Ok(response);
        });

        return group;
    }
}
```

Could someone elaborate on the security concerns on returning the claims? 

I'm also confused because I chose ASP.NET Core Identity based on [this recommendation](https://learn.microsoft.com/en-us/aspnet/core/security/how-to-choose-identity-solution?view=aspnetcore-8.0#determine-if-an-oidc-server-is-needed). It says there:

`
For apps that don't share logins with other apps, the simplest way to quickly secure an app is to use the built-in ASP.NET Core Identity provider. Otherwise, an OIDC server provided by a third-party identity management solution is needed.
`

I have **my own Blazor WASM** app that talks to **my own ASP.NET Core API**, and nothing else. 

What's the right way to make the claims a key part of this scenario?

**Comment by augustevn** [comment]
> I also stumbled into this today. It didn't break my app, since I was just trying to integrate my Blazor WASM app with the new identity endpoints as described in [this official doc](https://learn.microsoft.com/aspnet/core/blazor/security/webassembly/standalone-with-identity).
> 
> However, after completing the work, I realized the identity endpoints are not useful other than to register and authenticate a user.
> 
> And, the first thing you want to do after authenticating the user is check which permissions he has to decide how much to allow him to do on both the server and client side.
> 
> Without the claims, you can't tell the most basic thing: is he an admin? Can he modify things on my site?
> 
> I understand that this is not OIDC, so I would never be able to set scopes, which is the next thing I would want. But I must be able to at least check basic claims like the role the user belongs to.
> 
> To unblock, I implemented my own Identity endpoint group by extending the built-in ones:
> 
> ```cs
> public static class IdentityEndpoints
> {
>     public static RouteGroupBuilder MapIdentityEndpoints(this IEndpointRouteBuilder routes)
>     {
>         var group = routes.MapGroup("/identity")
>                           .WithTags("Identity");
> 
>         group.MapIdentityApi<GameStoreUser>();
> 
>         group.MapPost("/logout", async (ClaimsPrincipal user, SignInManager<GameStoreUser> signInManager) =>
>         {
>             await signInManager.SignOutAsync();
>             return TypedResults.Ok();
>         });
> 
>         group.MapGet("/manage/infoWithClaims", async Task<Results<Ok<InfoWithClaimsResponse>, ValidationProblem, NotFound>>
>             (ClaimsPrincipal claimsPrincipal, UserManager<GameStoreUser> userManager) =>
>         {
>             if (await userManager.GetUserAsync(claimsPrincipal) is not { } user)
>             {
>                 return TypedResults.NotFound();
>             }
> 
>             var email = await userManager.GetEmailAsync(user) ?? throw new NotSupportedException("Users must have an email.");
> 
>             var response = new InfoWithClaimsResponse(
>                 email,
>                 await userManager.IsEmailConfirmedAsync(user),
>                 claimsPrincipal.Claims.ToDictionary(c => c.Type, c => c.Value));
> 
>             return TypedResults.Ok(response);
>         });
> 
>         return group;
>     }
> }
> ```
> 
> Could someone elaborate on the security concerns on returning the claims?
> 
> I'm also confused because I chose ASP.NET Core Identity based on [this recommendation](https://learn.microsoft.com/en-us/aspnet/core/security/how-to-choose-identity-solution?view=aspnetcore-8.0#determine-if-an-oidc-server-is-needed). It says there:
> 
> `For apps that don't share logins with other apps, the simplest way to quickly secure an app is to use the built-in ASP.NET Core Identity provider. Otherwise, an OIDC server provided by a third-party identity management solution is needed.`
> 
> I have **my own Blazor WASM** app that talks to **my own ASP.NET Core API**, and nothing else.
> 
> What's the right way to make the claims a key part of this scenario?

I'm not in a position to answer security best practices questions, I assume exposing that "SecurityStamp" wasn't ideal. If I'm not mistaking, the Azure AD approach also exposes claims so...

By now, I also found out that there is no "token-rotation" after refresh token use which means any non-expired refresh token could still be re-used for a week+-. To be fair, they clearly stated this to be for simple scenario's and not to be(come) a token provider.

@julioct Interesting approach, do you prefer checking for an empty ClaimsPrincipal over authorizing the endpoint?

```
        identityGroup.MapGet("/user-info", (ClaimsPrincipal user) =>
        {
            int.TryParse(user.FindFirstValue(CustomClaimTypes.Permissions), out int permissions);
            
            return Results.Ok(new UserInfoResponse
            {
                Id = user.FindFirstValue(ClaimTypes.NameIdentifier) ?? string.Empty,
                Email = user.FindFirstValue(ClaimTypes.Email) ?? string.Empty,
                Permissions = permissions,
                Roles = user.FindAll(ClaimTypes.Role).Select(claim => claim.Value)
            });
        }).RequireAuthorization();
```

**Comment by julioct** [comment]
> I'm not in a position to answer security best practices questions, I assume exposing that "SecurityStamp" wasn't ideal. If I'm not mistaking, the Azure AD approach also exposes claims. I'm open for a better approach with or without recrafting the Blazor Identity components that expect those claims for the roles etc.
> 
> By now, I also found out that there is no "token-rotation" after refresh token use which means any non-expired refresh token could still be re-used for a week+-.
> 
> However, they stated this to be for simple scenario's for devices that can't handle cookies . And that is is not meant to be(come) a token server. I had hopes for this to be(come) the alternative or the preferred way over a custom JWT implementation.
> 
> @julioct Interesting approach, do you prefer checking for an empty ClaimsPrincipal over authorizing the endpoint?
> 
> ```cs
>     private static void MapIdentityEndpoints(this IEndpointRouteBuilder routeBuilder)
>     {
>         var identityGroup = routeBuilder
>             .MapGroup(ApiRoutes.Auth)
>             .WithTags(nameof(ApiRoutes.Auth));
>             
>         identityGroup.MapIdentityApi<User>();
> 
>         identityGroup.MapGet("/user-info", (ClaimsPrincipal user) =>
>         {
>             int.TryParse(user.FindFirstValue(CustomClaimTypes.Permissions), out int permissions);
>             
>             return Results.Ok(new UserInfoResponse
>             {
>                 Id = user.FindFirstValue(ClaimTypes.NameIdentifier) ?? string.Empty,
>                 Email = user.FindFirstValue(ClaimTypes.Email) ?? string.Empty,
>                 Permissions = permissions,
>                 Roles = user.FindAll(ClaimTypes.Role).Select(claim => claim.Value),
>                 Claims = user.Claims.ToDictionary(claim => claim.Type, claim => claim.Value)
>             });
>         }).RequireAuthorization();
>     }
> ```

Ohh I just copied the implementation I found in the ASP.NET Core repo just before they removed the Claims property from the /manage/info endpoint.

**Comment by lcetinapta** [comment]
I agree to this, there is no way of getting my user claims from the api endpoints anymore.

**Comment by carlfranklin** [comment]
I got part of the way there with the following solution, but it only gives me the roles during pre-rendering. Once we're in WASM land, the roles disappear. This code assumes we're using the Visual Studio Blazor Web App template with Individual accounts, WebAssembly interactive mode, and Global interactivity location. Also it assumes you've inserted .AddRoles<IdentityRole>() in Program.cs:

```
builder.Services.AddIdentityCore<ApplicationUser>(options => 
        options.SignIn.RequireConfirmedAccount = true)
    .AddRoles<IdentityRole>()
    .AddEntityFrameworkStores<ApplicationDbContext>()
    .AddSignInManager()
    .AddDefaultTokenProviders();
```

The first step is to modify UserInfo to support roles:

```
    public class UserInfo
    {
        public required string UserName { get; set; } // <= added
        public required string UserId { get; set; }
        public required string Email { get; set; }
        public string? Role { get; set; } // <= added
    }
```
Step 2 is to modify PersistingServerAuthenticationStateProvider on the server:

```
    internal sealed class PersistingServerAuthenticationStateProvider : ServerAuthenticationStateProvider, IDisposable
    {
        private readonly PersistentComponentState state;
        private readonly IdentityOptions options;

        private readonly PersistingComponentStateSubscription subscription;

        private Task<AuthenticationState>? authenticationStateTask;

        public PersistingServerAuthenticationStateProvider(
            PersistentComponentState persistentComponentState,
            IOptions<IdentityOptions> optionsAccessor)
        {
            state = persistentComponentState;
            options = optionsAccessor.Value;

            AuthenticationStateChanged += OnAuthenticationStateChanged;
            subscription = state.RegisterOnPersisting(OnPersistingAsync, RenderMode.InteractiveWebAssembly);
        }

        private void OnAuthenticationStateChanged(Task<AuthenticationState> task)
        {
            authenticationStateTask = task;
        }

        private async Task OnPersistingAsync()
        {
            if (authenticationStateTask is null)
            {
                throw new UnreachableException($"Authentication state not set in {nameof(OnPersistingAsync)}().");
            }

            var authenticationState = await authenticationStateTask;
            var principal = authenticationState.User;

            if (principal.Identity?.IsAuthenticated == true)
            {
                var userId = principal.FindFirst(options.ClaimsIdentity.UserIdClaimType)?.Value;
                var email = principal.FindFirst(options.ClaimsIdentity.EmailClaimType)?.Value;
                var name = principal.FindFirst(options.ClaimsIdentity.UserNameClaimType)?.Value;
                var role = principal.FindFirst(options.ClaimsIdentity.RoleClaimType)?.Value;

                if (userId != null && email != null)
                {
                    state.PersistAsJson(nameof(UserInfo), new UserInfo
                    {
                        UserId = userId,
                        Email = email,
                        UserName = name ?? email,
                        Role = role
                    });
                }
            }
        }

        public void Dispose()
        {
            subscription.Dispose();
            AuthenticationStateChanged -= OnAuthenticationStateChanged;
        }
    }
```
Step 3 is to modify PersistentAuthenticationStateProvider on the client:

```
internal class PersistentAuthenticationStateProvider : AuthenticationStateProvider
{
    private static readonly Task<AuthenticationState> defaultUnauthenticatedTask =
        Task.FromResult(new AuthenticationState(new ClaimsPrincipal(new ClaimsIdentity())));

    private readonly Task<AuthenticationState> authenticationStateTask = defaultUnauthenticatedTask;

    private readonly PersistentComponentState persistentState;

    public PersistentAuthenticationStateProvider(PersistentComponentState state)
    {
        if (!state.TryTakeFromJson<UserInfo>(nameof(UserInfo), out var userInfo) || userInfo is null)
        {
            return;
        }
        persistentState = state;

        Claim[] claims = [
            new Claim(ClaimTypes.NameIdentifier, userInfo.UserId),
            new Claim(ClaimTypes.Name, userInfo.Email),
            new Claim(ClaimTypes.Role, userInfo.Role),
            new Claim(ClaimTypes.Email, userInfo.Email) ];

        authenticationStateTask = Task.FromResult(
            new AuthenticationState(new ClaimsPrincipal(new ClaimsIdentity(claims,
                authenticationType: nameof(PersistentAuthenticationStateProvider)))));
    }

    public override Task<AuthenticationState> GetAuthenticationStateAsync()
    {
        if (persistentState == null)
            return defaultUnauthenticatedTask;

        if (!persistentState.TryTakeFromJson<UserInfo>(nameof(UserInfo), out var userInfo) || userInfo is null)
            return defaultUnauthenticatedTask;

        Console.Write("Foo");

        IEnumerable<Claim> EnumerateClaims()
        {
            yield return new(ClaimTypes.NameIdentifier, userInfo.UserId);
            yield return new(ClaimTypes.Name, userInfo.UserName); 
            yield return new(ClaimTypes.Email, userInfo.Email);
            if (userInfo.Role is string role)
            {
                yield return new(ClaimTypes.Role, role);
            }
        }

        return Task.FromResult(
            new AuthenticationState(new ClaimsPrincipal(new ClaimsIdentity(EnumerateClaims(),
                authenticationType: nameof(PersistentAuthenticationStateProvider)))));
    }
}
```
Like I said, the roles are there during pre-rendering, but the disappear at the first WASM rendering.

**Comment by carlfranklin** [comment]
I've created a stopgap solution: a replacement for AuthorizeView that supports Roles in .NET 8 Blazor WebAssembly apps. All the requirements are laid out in the repo, which can be found at https://github.com/carlfranklin/BlazorAuthorizeRoleView

Comments welcome.
